test/test_producer.py

```diff
@@ -40,3 +41,39 @@ def partitions(topic):
         topic = b"test-topic"
         producer.send_messages(topic, b'hi')
         assert client.send_produce_request.called
+
+    def test_producer_async_queue_overfilled_batch_send(self):
+        queue_size = 2
+        producer = Producer(MagicMock(), batch_send=True, maxsize=queue_size)
+
+        topic = b'test-topic'
+        partition = 0
+        message = b'test-message'
+
+        with self.assertRaises(Full):
+            message_list = [message] * (queue_size + 1)
+            producer.send_messages(topic, partition, *message_list)
+
+    def test_producer_async_queue_overfilled(self):
+        queue_size = 2
+        producer = Producer(MagicMock(), async=True, maxsize=queue_size)
+
+        topic = b'test-topic'
+        partition = 0
+        message = b'test-message'
+
+        with self.assertRaises(Full):
+            message_list = [message] * (queue_size + 1)
+            producer.send_messages(topic, partition, *message_list)
+
+    def test_producer_async_queue_normal(self):
+        queue_size = 4
+        producer = Producer(MagicMock(), async=True, maxsize=queue_size)
+
+        topic = b'test-topic'
+        partition = 0
+        message = b'test-message'
+
+        message_list = [message] * (queue_size / 2 + 1)
+        resp = producer.send_messages(topic, partition, *message_list)
+        self.assertEqual(type(resp), list)
```

Comment:
I think `len(resp) == len(message_list)`. Could you assert on that?

What's the state of the queue at the end here? Could you assert that as well?